datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -181,24 +189,19 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
             expr = expr.rewrite(&mut Canonicalizer::new()).data()?
         }
 
-        // TODO iterate until no changes are made during rewrite
-        // (evaluating constants can enable new simplifications and
-        // simplifications can enable new constant evaluation)
-        // https://github.com/apache/datafusion/issues/1160
-        expr.rewrite(&mut const_evaluator)
-            .data()?
-            .rewrite(&mut simplifier)
-            .data()?
-            .rewrite(&mut guarantee_rewriter)
-            .data()?
-            // run both passes twice to try an minimize simplifications that we missed
-            .rewrite(&mut const_evaluator)
-            .data()?
-            .rewrite(&mut simplifier)
-            .data()?
-            // shorten inlist should be started after other inlist rules are applied
-            .rewrite(&mut shorten_in_list_simplifier)
-            .data()
+        let mut num_iterations = 0;
+        loop {
+            let Transformed { data, transformed, .. } = expr
+                .rewrite(&mut const_evaluator)?
+                .transform_data(|expr| expr.rewrite(&mut simplifier))?
+                .transform_data(|expr| expr.rewrite(&mut guarantee_rewriter))?
```

Comment:
`guarantee_rewriter` and `shorten_in_list_simplifier` only ran once in the previous version of this code. Should we continue only running them once here?

Reply:
Yeah, that would probably be a good idea

Reply:
> Yeah, that would probably be a good idea

I can put `shorten_in_list_simplifier` at the end of the loop, since that's where it was previously. I am unsure where `guarantee_rewriter` is supposed to go as it previously ran once inbetween the simplifier/constant evaluator passes.